### PR TITLE
Integration specs for views and fix n+1 problems

### PR DESCRIPTION
**Pull request by Obimbo07**
- Made sure that the N+1 problem is solved when fetching all posts and their comments for a user.
Here is the logs before:
image
Here is the logs after: 
image

- Used Capybara to write integration tests for each view.

- [ ] User index page:

I can see the username of all other users.
I can see the profile picture for each user.
I can see the number of posts each user has written.
When I click on a user, I am redirected to that user's show page.

- [ ] User show page:

I can see the user's profile picture.
I can see the user's username.
I can see the number of posts the user has written.
I can see the user's bio.
I can see the user's first 3 posts.
I can see a button that lets me view all of a user's posts.
When I click a user's post, it redirects me to that post's show page.
When I click to see all posts, it redirects me to the user's post's index page.

- [ ] User post index page:

I can see the user's profile picture.
I can see the user's username.
I can see the number of posts the user has written.
I can see a post's title.
I can see some of the post's body.
I can see the first comments on a post.
I can see how many comments a post has.
I can see how many likes a post has.
I can see a section for pagination if there are more posts than fit on the view.
When I click on a post, it redirects me to that post's show page.

- [ ] Post show page:

I can see the post's title.
I can see who wrote the post.
I can see how many comments it has.
I can see how many likes it has.
I can see the post body.
I can see the username of each commentor.
I can see the comment each commentor left.

### CODE REVIEW

File: spec/features/user_show_spec.rb [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+    expect(page).to have_selector('.post', count: 2)
+    expect(page).to have_link('Post', href: user_post_path(user, post1))
+    expect(page).to have_link('Post', href: user_post_path(user, post2))
+    expect(page).to have_content(post1.title)
+    expect(page).to have_content(post2.title)
+    expect(page).to have_css('.count', text: "Comments: #{post1.comments.count} | Likes: #{post1.likes.count}")
+    expect(page).to have_css('.count', text: "Comments: #{post2.comments.count} | Likes: #{post2.likes.count}")
+
+    expect(page).to have_link('See all posts', href: user_posts_path(user))
```
Comment (rloterh) [inline spec/features/user_show_spec.rb]: - [ ] The test to assert the scenario, **_I can see the user's first 3 posts_** for _user show_ view integration spec is missing! Please ensure to improve upon this aspect.

File: spec/features/post_index_spec.rb [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    expect(page).to have_content(user.posts.first.text)
+    expect(page).to have_css('.count', text: "Comments: #{user.posts.first.comments.count}")
+    expect(page).to have_css('.count', text: "Likes: #{user.posts.first.likes.count}")
+
+    expect(page).to have_selector('.comments', count: 1)
+    user.posts.first.recent_comments.each do |comment|
+      expect(page).to have_content("#{comment.user.name} : #{comment.text}")
+    end
+
+    expect(page).to have_link('Pagination', href: user_posts_path(user))
```
Comment (rloterh) [inline spec/features/post_index_spec.rb]: - [ ] The test to assert the scenario, **_I can see a post's title_** for _post index_ view integration spec is missing! Please ensure to improve upon this aspect.

File: spec/features/post_show_spec.rb [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+    expect(page).to have_selector('.container')
+    expect(page).to have_selector('.post')
+    expect(page).to have_content("Post: #{post.title}")
+    expect(page).to have_css('.count', text: "Comments: #{post.comments.count}")
+    expect(page).to have_css('.count', text: "Likes: #{post.likes.count}")
+    expect(page).to have_content(post.text)
+
+    expect(page).to have_selector('.comments')
+    post.comments.each do |comment|
+      expect(page).to have_selector('.comment', text: "#{comment.user.name} : #{comment.text}")
```
Comment (rloterh) [inline spec/features/post_show_spec.rb]: - [ ] The test to assert the scenario, **_I can see who wrote the post_** for _post show_ view integration spec is missing! Please ensure to improve upon this aspect.

File: spec/features/user_index_spec.rb [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+RSpec.feature 'Users Index View', type: :feature do
+  scenario 'User views users index page' do
+    user1 = User.create(name: 'Alice', photo: 'https://example.com/alice.jpg')
+    user1.posts.create(title: 'Post 1', text: 'First post by Alice')
+    user1.posts.create(title: 'Post 2', text: 'Second post by Alice')
+    user2 = User.create(name: 'Bob', photo: 'https://example.com/bob.jpg')
+    user2.posts.create(title: 'Post 3', text: 'First post by Bob')
+    visit users_path
+    expect(page).to have_selector('.container')
+    expect(page).to have_selector('.user_info', count: 4)
```
Comment (rloterh) [inline spec/features/user_index_spec.rb:11]: - [ ] A user index view spec is failing. Let's ensure all spec scenarios are passing!

**Failing spec:**
image

**Review by rloterh — CHANGES_REQUESTED** [review]
Hi Obimbo, 

You did extremely great on this project! 👏 👏 
There are however a few issues that you still need to work on to go to the next project, but you are almost there! 
🏁 🚦  ⛪ 🚜  🏃‍♀️  🌴  🏫 🌳 🏘️ 

## Highlights 👍  
- N + 1 problem addressed!
- ReadMe looks great!

## Required Changes ♻️
_Kindly check the inline comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough 
Cheers and Happy coding!👏 💻 🍷

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think the code review was unfair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Comment by Reem-lab** [comment]
Hello @Obimbo07 could you check your slack?

**Comment by rloterh** [comment]
Hi @Obimbo07

Robocop linter checks are failing, can you ensure all linters are passing ASAP?

Standing by...cheers!

**Review by rloterh — COMMENTED** [review]
Hi Austin,
 
 While you made a great effort in this project, unfortunately, I cannot proceed to review your code.
 
 ## Invalid Code Review Request
You have submitted a project, whereby some Rubocop linter checks are failing on your Pull Request. Kindly ensure that **all** checks are passing.

**Failing Checks:**
image
 
Your Code Review Request will be marked as invalid in your Dashboard, so please submit a new one once you are ready 🙏
 
 Cheers and Happy coding!👏👏👏
 
 Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
 
 **_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

--------

_Invalid Code Review Request **does not count** into the code reviews limit._

File: spec/features/user_index_spec.rb (hunk 4 above)
Comment (thecodechaser) [inline spec/features/user_index_spec.rb:11]: - [ ] As mentioned by the previous reviewer the below test is failing, Please make sure that all tests pass without any errors.

image

**Review by thecodechaser — CHANGES_REQUESTED** [review]
## STATUS: CHANGES REQUIRED ♻️ 

## Hi @Obimbo07, and Team👋, 

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.

## To Highlight 🎉
- Good job fixing n+1 problem. ✔️ 
- Good job with integration tests. ✔️
- Gitflow is followed. :heavy_check_mark:
- Professional README file. :heavy_check_mark:
- No linter errors. :heavy_check_mark:
- PR has a good title and summary. ✔️



 ## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/features/user_index_spec.rb (hunk 4 above)
Comment (Obimbo07) [reply spec/features/user_index_spec.rb:11]: Hello,

When I submit my tests they pass but I don't know why on reviewers side it doesn't. The previous reviewer told me to submit this video in the comment section.

https://github.com/Obimbo07/tdn/assets/118368849/bb2d1a69-9fac-4adc-aa29-f4a2c6ab8e0e

**Review by thecodechaser — APPROVED** [review]
## STATUS: APPROVED 🟢

## Hi @Obimbo07, and Team👋, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## To Highlight 🎉
- Good job fixing n+1 problem. ✔️
- Good job with integration tests. ✔️
- Gitflow is followed. :heavy_check_mark:
- Professional README file. :heavy_check_mark:
- No linter errors. :heavy_check_mark:
- PR has a good title and summary. ✔️



Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**